importers/aws/aws.go

```diff
@@ -0,0 +1,704 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package aws implements AWS repository importer.
+package aws
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/golang/glog"
+	"github.com/google/hashr/core/hashr"
+	"github.com/google/hashr/importers/common"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	RepoName     = "AWS"
+	buildTimeout = 1800 // 30 minutes
+)
+
+// AWS API clients and global configuration vars.
+var (
+	ec2Client  *ec2.Client
+	s3Client   *s3.Client
+	bucketName string
+	sshUser    string
+)
+
+// instanceMap stores the state of the available EC2 HashR worker instances.
+//
+// If an EC2 instance is currently processing an AMI, instanceMap is set where
+// instance ID is the key and the AMI ID being processed is the value.
+//
+// If a key (instance ID) does not exist, it means it is not in use.
+var instanceMap map[string]string
+
+func init() {
+	instanceMap = make(map[string]string)
+}
+
+// image holds data related to AWS image.
+type image struct {
+	localImage      types.Image
+	sourceImage     types.Image
+	localTarGzPath  string
+	remoteTarGzPath string
+	quickSha256Hash string
+	instance        types.Instance
+	regionName      string
+	volumeId        string
+	device          string
+	sshClient       *ssh.Client
+}
+
+// Preprocess creates tar.gz file from an image, copies to local storage, and extracts it.
+func (i *image) Preprocess() (string, error) {
+	var err error
+
```

Comment:
Instead of creating context.TODO() multiple times, can you do it once here in this function and pass to all the functions that need it? According to https://pkg.go.dev/context this is the way. At some point in the future I'll change the importer interface so that we can pass context directly to all the functions that needs to satisfy it (e.g. Preprocess)

Reply:
Fixed